### PR TITLE
Fix #549: GC-aware tail-call optimization for allocating functions

### LINKED ISSUE #549: GC-aware tail-call optimization for allocating functions
## Summary

v0.0.126 (#517) shipped WASM `return_call` emission for tail-position calls but disabled the optimization for allocating functions (functions whose body translation set `ctx.needs_alloc = True`).  This issue tracks the follow-up: extending TCO to allocating functions by emitting GC-shadow-stack cleanup before the `return_call`.

## Why allocating functions were excluded

WASM `return_call $foo` discards the current frame and jumps to `$foo`.  In Vera's GC model, that means the **GC epilogue never runs** — the saved `$gc_sp` is not restored, and the shadow-stack pointer slots pushed by the prologue stay live.  Each tail-call iteration thus leaks shadow-stack slots; eventually `$gc_sp` passes the worklist boundary and the next `$alloc` traps with out-of-bounds.  Strictly *worse* for some recursion shapes than the pre-fix "call stack exhausted" trap (silent shadow-stack corruption rather than a clean trap).

The post-process at the end of `_compile_fn` (`vera/codegen/functions.py`) reverts every `return_call` → `call` when `ctx.needs_alloc` is True, paying the WASM frame cost in exchange for correct shadow-stack management.

## What needs to happen

Emit a "GC tail-call epilogue" before each `return_call` in an allocating function:

```wat
;; Restore $gc_sp to entry value BEFORE evaluating call args.
;; Args may allocate; those allocations push from the entry-level
;; baseline, so the callee's prologue sees a clean shadow stack.
local.get $gc_sp_save
global.set $gc_sp

;; Now evaluate args — any allocation here pushes from clean state.
... arg expressions ...

return_call $foo
```

The semantics:
- At entry: `$gc_sp` is at value S0 (caller's pre-call value).
- Prologue: save S0 to local `$gc_sp_save`, push pointer params, `$gc_sp` is now S1.
- ... function body executes, may allocate further ...
- At return_call: restore `$gc_sp` to S0 BEFORE arg evaluation.
- Args evaluate from S0 baseline; any allocations push correctly.
- `return_call` jumps to callee; callee's prologue saves the new gc_sp baseline.

## Implementation sketch

In `_compile_fn` (`vera/codegen/functions.py`):
1. Track which body instructions are `return_call $X` emissions.
2. When `needs_alloc` is True, instead of reverting to plain `call`, prepend the GC-restore-prologue to each `return_call`.
3. Need to ensure the args don't reference locals that were freed by the gc_sp restore — they shouldn't, locals are independent of `$gc_sp`.

## Tests

Once shipped:
- A tail-recursive allocating function (e.g. building an ADT each iteration) should run for 1M+ iterations without trapping.
- Structural assertion: the `return_call` site is preceded by the GC-restore sequence.
- Behavioural: equivalent of `test_tail_recursive_iteration_succeeds_at_1m` but with allocation.

## Out of scope

- TCO for closures called via `call_indirect` (would need `return_call_indirect`); separate issue if there's user demand.

## Notes

Discovered as a deliberate scoping decision in v0.0.126 (#517).  The current behaviour is correct (allocating functions get plain `call`, paying the stack cost); this issue extends correctness by also flattening the stack.  Probably 30-60 minutes of focused work given the existing prologue/epilogue infrastructure.

### PR DESCRIPTION
**Pull request by aallan**
## Summary

- **Replaces the "allocating fn ⇒ revert `return_call` → `call`" fallback with a GC-aware patch.** The post-process in `vera/codegen/functions.py::_compile_fn` now PREPENDS `local.get $gc_sp_save; global.set $gc_sp` immediately before each `return_call` site in an allocating fn, restoring the shadow-stack pointer to the caller's entry baseline so the callee saves a clean new baseline. Per-iteration shadow-stack usage stays bounded at `caller's entry + n_arg_roots` regardless of iteration count.
- **Args transfer atomically.** At the `return_call` site the args are already on the WASM operand stack; the two-instruction restore touches only the `$gc_sp` *global*, not the operand stack, so the patch is non-interfering — `return_call` still consumes args and atomically transfers control.
- **Postcondition revert preserved.** Functions with a non-trivial runtime postcondition still revert `return_call` → plain `call` (the post-check needs to run after each call; `return_call` would skip it). Dispatch precedence: `post_instrs` revert > `needs_alloc` GC-aware patch > untouched.
- **Closes the last open bug in `KNOWN_ISSUES.md` bug-tracker section apart from #578** — the bug list will be empty once #578 lands.

## Test plan

- [x] `pytest tests/` — 3,862 passed, 14 skipped, 16 deselected
- [x] `mypy vera/` — clean
- [x] `python scripts/check_conformance.py` — 86/86
- [x] `python scripts/check_examples.py` — 34/34
- [x] `python scripts/check_doc_counts.py` — consistent
- [x] `python scripts/check_version_sync.py` — 0.0.154 across 6 files
- [x] `python scripts/check_site_assets.py` — up-to-date
- [x] `python scripts/check_limitations_sync.py` — consistent
- [x] Pre-commit hooks — all 18 passed including conformance suite + example validation
- [x] Stress test sample (`pytest tests/test_stress.py::test_tco_with_allocation_1m_iterations -m stress`) — 1M iterations under both default and eager GC in ~190ms each

## Behavioural impact

Pre-#549, an allocating tail-recursive function like:

```vera
private fn loop(@Int, @Int -> @Int)
  requires(@Int.1 >= 0) ensures(true) decreases(@Int.1) effects(pure)
{
  if @Int.1 == 0 then { @Int.0 }
  else {
    let @Array<Int> = [@Int.0, @Int.1];
    loop(@Int.1 - 1, @Int.0 + array_length(@Array<Int>.0))
  }
}
```

trapped with `stack_exhausted` at ~30K iterations (the WASM call-stack limit) because the post-process reverted `return_call` → plain `call`. Post-#549 the same function runs in constant stack space at 1M+ iterations — the per-iteration `$gc_sp` restore keeps the shadow stack flat across the entire run.

## Test changes

- `tests/test_codegen.py::TestTailCallOptimization517` — renamed `test_allocating_function_falls_back_to_plain_call` to `test_allocating_function_uses_gc_aware_tco_549` and inverted its assertions: it now verifies `return_call $foo` is emitted AND every such site is preceded by `local.get <N>; global.set $gc_sp`. Added a sibling test `test_allocating_function_with_postcondition_still_reverts` to pin the postcondition-revert precedence.
- `tests/test_codegen.py::TestGCShadowStackOverflow::test_shadow_stack_overflow_traps` — rewritten to use a non-tail-recursive shape (recursive call wrapped in `array_append`) so the overflow guard still genuinely trips post-#549.
- `tests/test_stress.py::test_deep_tail_recursion_with_allocating_arg` — body switched from string-pool literals (didn't actually set `needs_alloc`) to genuine array allocation, so the test now actually exercises #549's path.
- `tests/test_stress.py::test_tco_with_allocation_1m_iterations` — new 1M-iteration companion parametrised over default-GC and eager-GC modes (~190ms wall-clock in both).

## Doc sweep

- `KNOWN_ISSUES.md` — removed the #549 bug row.
- `SKILL.md` (+ regenerated `docs/SKILL.md`) — removed the "tail-call optimization disabled for allocating functions" row from the bug table; updated the narrative paragraph below it to describe both #517 and #549 as live optimisations.
- `vera/wasm/calls.py` — comment near `_translate_call` updated to describe both branches of the post-process (GC-aware patch vs postcondition revert).
- `vera/codegen/api.py` — the `stack_exhausted` runtime trap-fix message updated: drops the "allocating functions are an exception" paragraph, adds a brief mention of the postcondition-revert as the remaining exception.
- Version bumped to v0.0.154 across `pyproject.toml`, `vera/__init__.py`, `uv.lock`, `docs/index.html`, `README.md`.
- `CHANGELOG.md` v0.0.154 entry added with the design rationale, plus `HISTORY.md` row.
- `TESTING.md`, `ROADMAP.md`, `README.md` — test counts updated to 3,892 (was 3,889).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Restored GC‑aware tail‑call optimisation for allocating functions so shadow‑stack/GC state is preserved across recursive tail calls while retaining correct postcondition behaviour.

* **Tests**
  * Added and expanded regression and stress suites (including a 1,000,000‑iteration stress case) to validate GC‑aware TCO, overflow/trap behaviour and updated stress parametrisation and expectations.

* **Documentation**
  * Bumped project version; updated changelogs, README, HISTORY, ROADMAP, TESTING and runtime guidance; removed the resolved known‑issue entry.

[Review Change Stack](https://app.coderabbit.ai/change-stack/aallan/vera/pull/671)

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -298,6 +298,7 @@ Stage 12 opens on the morning v0.0.138 shipped: the residual GC-rooting bug in #
 | v0.0.151 | 12 May | **Walker-completeness audit with pre-commit enforcement** ([#597](https://github.com/aallan/vera/issues/597)). |
 | v0.0.152 | 13 May | **Stress-test harness for scale-dependent regression coverage** ([#596](https://github.com/aallan/vera/issues/596)). |
 | v0.0.153 | 13 May | **SMT translator covers `FloatLit` / `IndexExpr` / `ArrayLit` in contracts** ([#667](https://github.com/aallan/vera/issues/667)). |
+| v0.0.154 | 13 May | **GC-aware tail-call optimization for allocating functions** ([#549](https://github.com/aallan/vera/issues/549)). |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md:301]: @coderabbitai fixed in ae25d84 — the footer now reads "154 tagged releases" so it matches the v0.0.154 row and README's release count.

File: tests/test_codegen.py [hunk 2]
```diff
@@ -2057,20 +2080,99 @@ def test_allocating_function_falls_back_to_plain_call(self) -> None:
...
+        # plain call (so the postcondition check actually runs).
         assert "call $build" in build_body
         assert "return_call $build" not in build_body, (
-            f"Allocating function `build` emitted return_call, "
-            f"which would leak shadow-stack slots.  Post-process "
-            f"should have reverted to plain call.  Body:\n"
-            f"{build_body}"
+            f"build has a runtime postcondition; return_call would "
+            f"skip it. Post-process should have reverted to plain "
+            f"call. Body:\n{build_body}"
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:2336]: @coderabbitai fixed in ae25d84 — added a structural assertion that no `local.get N; global.set $gc_sp` pair precedes any plain `call $build` site, pinning the dispatch precedence as mutually exclusive (post_instrs revert and GC-aware patch are alternatives, not additive). Can't forbid the two-instruction sequence outright because the GC epilogue at function-end legitimately emits it.

File: tests/test_codegen.py [hunk 3]
```diff
@@ -2057,20 +2080,99 @@ def test_allocating_function_falls_back_to_plain_call(self) -> None:
...
+            assert prev1 == "global.set $gc_sp", (
+                f"Expected 'global.set $gc_sp' immediately before "
+                f"return_call at line {idx}, got: {prev1!r}. "
+                f"Body:\n{build_body}"
+            )
+            assert prev2.startswith("local.get "), (
+                f"Expected 'local.get <N>' two lines before "
+                f"return_call at line {idx} (loading $gc_sp_save), "
+                f"got: {prev2!r}. Body:\n{build_body}"
+            )
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai fixed in ae25d84 — the test now parses the GC prologue (`global.get $gc_sp; local.set <N>`) at function entry to capture the exact `$gc_sp_save` local index, then asserts each `return_call $build` site is preceded by `local.get <that same N>; global.set $gc_sp`. A typo or off-by-one in the patch code that picked up an unrelated local would now fail.

File: vera/codegen/functions.py [hunk 4]
```diff
@@ -347,42 +347,87 @@ def _compile_fn(
...
+            # inheriting the leaked shadow-stack slots from this
+            # frame's arg-evaluation leg.  Args are already on the
+            # WASM operand stack at the return_call site; the
+            # restore touches only the ``$gc_sp`` global, not the
+            # operand stack, so args transfer atomically to the
+            # callee.  Pre-#549 this revert-to-plain-call path also
+            # fired for allocating fns; closing #549 lets allocating
+            # tail-recursive fns iterate indefinitely without
+            # unbounded shadow-stack growth.
+            assert gc_sp_save is not None
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/codegen/functions.py]: @coderabbitai CI lint failure resolved in ae25d84 via `# noqa: S101` on both sites, matching the existing project precedent in `vera/codegen/api.py` (5 sites use the same `assert ... # noqa: S101` pattern for mypy type-narrowing). Declined the `_error()` restructuring: both asserts are pure mypy narrowers inside `if ctx.needs_alloc:` blocks where `gc_sp_save` is provably non-None (it was allocated under the same condition a few lines up), so an `_error()` raise would be unreachable code. The `-O` strip concern is real in principle but the project doesn't run under `-O` and the failure mode (stripped assert → `local.set None` template → assemble error) is loud enough to debug.

File: TESTING.md [hunk 5]
```diff
@@ -58,15 +58,15 @@ python scripts/fix_allowlists.py --fix               # auto-fix stale allowlists
...
-| `test_stress.py` | 14 | 487 | Scale-dependent regression tests (#596) — `@pytest.mark.stress`, skipped by default.  8 logical tests × eager-GC lane parametrisation = 14 test instances.  10K `array_map`, 5K nested-array `array_map`, 1K-deep tail recursion with allocating arg, 20×20 nested array-fold-of-array-fold, 100K `array_fold`, 10K String allocations, 1K `State<Int>` get/put cycles, 10K `IO.print` calls.  Pins #570 / #515 / #593 / #549 / #487 / #348 / #573 regression coverage |
+| `test_stress.py` | 16 | 551 | Scale-dependent regression tests (#596) — `@pytest.mark.stress`, skipped by default.  9 logical tests × eager-GC lane parametrisation = 16 test instances.  10K `array_map`, 5K nested-array `array_map`, 1K-deep tail recursion with allocating arg, 1M-deep tail recursion with allocating arg (#549 GC-aware TCO), 20×20 nested array-fold-of-array-fold, 100K `array_fold`, 10K String allocations, 1K `State<Int>` get/put cycles, 10K `IO.print` calls.  Pins #570 / #515 / #593 / #549 / #487 / #348 / #573 regression coverage |
```
[the remark this answers is not in the document]
Comment (aallan) [reply TESTING.md]: @coderabbitai fixed in 846ce05 — synchronised stress totals across the document.

- `### The 8 initial test programs` → `### The 9 initial test programs`
- Added the missing **9. `test_tco_with_allocation_1m_iterations`** entry so the section actually enumerates what the header claims.
- Updated test #3's stale narrative — this PR switched its body from `let @String = "stress"` (string-pool, doesn't trigger `needs_alloc`) to `let @Array<Int> = [_, _]` (genuine heap alloc), and the asserted accumulator from 6,000 to 2,000.  Without this update the doc described a test that no longer exists.
- "Six of the eight tests" → "Seven of the nine tests" in the eager-GC paragraph.
- CLI examples and the Budget line updated from 14 instances / 8 logical to 16 instances / 9 logical.

File: tests/test_codegen.py [hunk 6]
```diff
@@ -2057,21 +2080,166 @@ def test_allocating_function_falls_back_to_plain_call(self) -> None:
...
+        gc_sp_save_local = prologue_set[len("local.set "):]
+        expected_preamble_get = f"local.get {gc_sp_save_local}"
+        # Find every `return_call $build` site and verify the two
+        # instructions immediately before it are the exact preamble:
+        #     local.get <gc_sp_save_local>
+        #     global.set $gc_sp
+        return_call_indices = [
+            i for i, line in enumerate(lines)
+            if "return_call $build" in line
+        ]
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai fixed in 846ce05 — switched all three #549 tests to boundary-safe `re.search` patterns. The trailing `\b` after `$build` correctly excludes a hypothetical `$build_helper`, and the leading `\b` before `call $build` correctly distinguishes from `return_call $build`. Same treatment applied to `$overflow` in `test_shadow_stack_overflow_traps`.

One amendment to the literal suggestion: dropped the leading `\b\(` from `r'\b\(func \$build\b'`. `\b` requires a word char on one side and `(` is preceded by whitespace or newline (both non-word) at WAT indentation depths, so that anchor wouldn't match indented `(func $build` lines. Kept the substantive trailing `\b` which is what blocks the `$build_anything` false-match.

Scope decline: did NOT extend the regex switchover to the pre-existing #517 tests in the file (lines 1899/1934/1949/1968/2016/2027 — `test_return_call_emitted_for_tail_position_call` and siblings). Those existed before this PR and changing them is out of scope for the #549 fix. Happy to do that sweep in a follow-up PR if useful — it'd be a straightforward mechanical pass with no behavioural change.

**Comment by aallan** [comment]
## PR self-review summary

Ran the `pr-review-toolkit:review-pr` workflow against this PR — three specialised agents in parallel: **code-reviewer**, **pr-test-analyzer**, **comment-analyzer**. Skipped type-design-analyzer (no new types) and silent-failure-hunter (no error-handling paths changed; the only new control flow is dispatch logic).

### code-reviewer — clean

> "No issues found at confidence ≥ 80. The PR is clean. The core fix is correct, well-commented, and well-tested. The dispatch precedence is explicitly pinned by a dedicated test. The prefix-preservation logic is defensible and the prefix-extraction idiom is correct."

Verified independently:
- Dispatch precedence (post_instrs revert > GC-aware patch > untouched) — correct.
- `gc_sp_save` pre-allocation at `functions.py:380-382` is index-safe; no conflict with later `gc_ret_ptr` / `gc_ret_len` / `gc_ret` locals.
- Prefix-preservation idiom `instr[: len(instr) - len(instr.lstrip())]` correct for WAT (spaces only, no tabs).
- Substring safety on `lstrip().startswith("return_call ")` confirmed — `return_call` only appears as instruction prefix in emission, never as comment or literal.
- Edge cases (`needs_alloc` with zero return_call sites; `post_instrs` without `needs_alloc`) handled correctly.

### pr-test-analyzer — one important gap, one suggestion

**Important** — no multi-branch `return_call` test (rating 8). The patch loop iterates every line and patches every site, but the single-branch fixture wouldn't catch a buggy implementation that bails after the first match (e.g. accidental `break`).

**Fixed in [`ae573d3`](https://github.com/aallan/vera/commit/ae573d3)** — added `test_allocating_function_gc_aware_tco_patches_both_branches`. Uses a `match` with two ADT arms each ending in `return_call $build`, asserts BOTH sites have the `local.get N; global.set $gc_sp` preamble AND that N matches the `$gc_sp_save` local captured by the prologue. A loop-early-exit bug would now fail loudly.

**Suggestion** — undocumented 2000-iteration figure in `test_shadow_stack_overflow_traps`. **Fixed**: inline comment documents the calibration (16K shadow stack / 12 bytes per frame ≈ 1,365 frames; 2000 = ~1.5× safety margin).

**Declined**:
- Mutual-recursion test with asymmetric `needs_alloc` (rating 8) — dispatch is per-function and each is independently pinned; cross-function GC composition is a deeper concern that doesn't sit in #549's scope. Worth a follow-up issue if the property matters.
- Zero-`return_call`-site test for allocating leaf fns (rating 6) — the edge case is benign (`gc_sp_save` is allocated but unused; no harm).

### comment-analyzer — four factual fixes

All four flagged in the [comment-analyzer report](https://github.com/aallan/vera/pull/671) and **fixed in [`ae573d3`](https://github.com/aallan/vera/commit/ae573d3)**:

| # | File | Fix |
|---|---|---|
| 1 | `tests/test_stress.py:248` | "~12 bytes per leaked frame ≈ 1300 iters" was wrong. `gc_shadow_push` writes one i32 = 4 bytes. The 12-byte figure came from a different fn elsewhere in the file. The `loop` here pushes 1 pointer per iter = 4 bytes; overflow threshold ≈ 4,096 iters. Also dropped the wall-clock "~200ms on 2026 MacBook Pro" claim (will rot). |
| 2 | `vera/codegen/functions.py:415` | Comment cited `operators.py` unqualified, but `vera/codegen/` and `vera/wasm/` are peer directories. Fully qualified to `vera/wasm/operators.py`. Added a #672 cross-reference (the broader fixup for inline-indent stamping). |
| 3 | `vera/codegen/functions.py:350` | Comment header said "Two sources of post-body work" but the actual dispatch is three-branch (revert / patch / untouched). Reframed as "Three outcomes (precedence: 1 > 2 > 3)" with the third explicitly named — the common non-allocating tail-recursion case. |
| 4 | `tests/test_codegen.py:2034` | Docstring said "sacrificing TCO for shadow-stack correctness". Inverted framing — pre-#549 the GC epilogue ran fine; what was sacrificed was WASM call-stack depth (1M iters trapped `call stack exhausted`). Reframed correctly. |

**Declined**:
- Adding "split into inner helper without postcondition + outer wrapper" workaround to `api.py`'s `stack_exhausted` trap message — too tangential for end-user-facing text.

### Validation

After `ae573d3`:
- pytest **3,863 passed**, 14 skipped, 16 stress-deselected (test count went 3,892 → 3,893 from the new multi-branch test)
- mypy clean (59 source files)
- ruff S rules clean
- doc-counts script consistent (3,893 tests across 32 files)
- All 18 pre-commit hooks green (conformance 86/86, examples 34/34, walker coverage, license, site assets, browser parity, etc.)

### Commit trail on this PR

| Commit | Purpose |
|---|---|
| `1df1e8b` | Initial #549 fix + tests + doc sweep + v0.0.154 release prep |
| `d7e54e4` | Preserve `return_call` leading whitespace when patching `$gc_sp` restore |
| `d78028e` | Add #672 (canonicalize WAT formatting) to ROADMAP CI tooling |
| `ae25d84` | Address CodeRabbit round-1 + fix lint failure |
| `846ce05` | Address CodeRabbit round-2 findings (TESTING.md sync + boundary-safe regex) |
| `ae573d3` | Address self-review findings (this comment) |

**Comment by aallan** [comment]
@coderabbitai full review
